core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java

```diff
@@ -561,6 +591,213 @@ public UpdateNode plan(Update node)
+        }
+
+        FieldReference reference = analysis.getRowIdField(mergeAnalysis.getTargetTable());
+        Symbol rowId = target.getFieldMappings().get(reference.getFieldIndex());
+        projectionAssignmentsBuilder.put(rowId, rowId.toSymbolReference());
+        Symbol mergeOutput = symbolAllocator.newSymbol("merge_row", rowType);
+        projectionAssignmentsBuilder.put(mergeOutput, caseExpression);
+
+        ProjectNode project = new ProjectNode(
+                idAllocator.getNextId(),
+                subPlan.getRoot(),
+                projectionAssignmentsBuilder.build());
+
+        // Projecting the writeRedistributionColumns, the rowId block, and the merge RowBlock
+        int expectedSize = redistributionColumns.size() + 1 + 1;
+        int actualSize = project.getOutputSymbols().size();
+        checkArgument(actualSize == expectedSize, "projectedSymbols should have size %s, but is %s", expectedSize, actualSize);
+
+        Table table = merge.getTargetTable();
+        TableHandle handle = analysis.getTableHandle(table);
+        TableMetadata tableMetadata = metadata.getTableMetadata(session, handle);
+
+        RowChangeParadigm paradigm = metadata.getRowChangeParadigm(session, handle);
+        Type rowIdType = analysis.getType(analysis.getRowIdField(table));
+        RowChangeProcessor rowChangeProcessor = createMergeProcessor(paradigm, tableMetadata, mergeAnalysis.getDataColumnHandles(), mergeAnalysis.getRedistributionColumnHandles(), rowIdType);
+        MergeTarget mergeTarget = new MergeTarget(handle, Optional.empty(), tableMetadata.getTable(), rowChangeProcessor);
+
+        ImmutableList.Builder<Symbol> columnSymbolsBuilder = ImmutableList.builder();
+        for (ColumnHandle columnHandle : mergeAnalysis.getDataColumnHandles()) {
+            int fieldIndex = requireNonNull(mergeAnalysis.getColumnHandleFieldNumbers().get(columnHandle), "Could not find field number for column handle");
+            columnSymbolsBuilder.add(target.getFieldMappings().get(fieldIndex));
+        }
+        List<Symbol> columnSymbols = columnSymbolsBuilder.build();
+
+        ImmutableList.Builder<Symbol> projectedSymbolsBuilder = ImmutableList.builder();
+        projectedSymbolsBuilder.addAll(columnSymbols);
+        Symbol operationSymbol = symbolAllocator.newSymbol("operation", INTEGER);
+        projectedSymbolsBuilder.add(operationSymbol);
+        projectedSymbolsBuilder.add(rowId);
+
+        List<Symbol> finalProjectedSymbols = projectedSymbolsBuilder.build();
+
+        Optional<PartitioningScheme> partitioningScheme = createPartitioningScheme(mergeAnalysis.getNewTableLayout(), columnSymbols, mergeAnalysis.getPartitionFunctionArgumentIndexes());
+        DeleteAndInsertNode deleteAndInsertNode = new DeleteAndInsertNode(
+                idAllocator.getNextId(),
+                project,
+                mergeTarget,
+                finalProjectedSymbols);
```

Comment:
My guess is that the `DeleteAndInsertNode` needs the `rowId` symbol and the "merge_row" symbol from the underlying projection. While `rowId` can be identified among the input symbols based on it's position in `DeleteAndInsertNode.outputs`, there is no way to identify the "merge_row" symbol. There's no guarantee that the symbol name will remain the same after the optimization phase, and no guarantee that the output order of the projection will remain unchanged.

If you want to hold on to these symbols, record them in the `DeleteAndInsertNode`. It should be also followed by a check in `ValidateDependenciesChecker`.

Reply:
Changed/added.